### PR TITLE
HHIOS-399 New Design for Map Layers

### PR DESCRIPTION
**Pull request by FatooAJ**
### Description
Our design team change the ui of the map layer sheet

### Open Tasks

- [x] change the UI of the map layer sheet
- [x] Allow user to turn on/off traffic (other buttons not yet working) 

### Infos for Reviewer
Screenshot 2024-11-14 at 2 29 50 PM
Screenshot 2024-11-14 at 2 29 40 PM

### CODE REVIEW

File: BackendService/Sources/BackendService/Toursprung/HudHudMapLayerStore.swift [hunk 1]
```diff
@@ -58,11 +65,71 @@ public class HudHudMapLayerStore: ObservableObject {
...
+
+    // MARK: Static Properties
+
+    public static let traffic = MapStyleOption(rawValue: 1 << 0)
+    public static let saved = MapStyleOption(rawValue: 1 << 1)
+    public static let streetView = MapStyleOption(rawValue: 1 << 2)
+    public static let road = MapStyleOption(rawValue: 1 << 3)
+
+    public static let all: MapStyleOption = [.traffic, .saved, .streetView, .road]
+    public static var allCases: [MapStyleOption] = [.traffic, .saved, .streetView, .road]
```
Comment (Patrick-Kladek) [inline BackendService/Sources/BackendService/Toursprung/HudHudMapLayerStore.swift:90]: Great idea with the OptionSet

File: BackendService/Sources/BackendService/Toursprung/HudHudMapLayerStore.swift [hunk 2]
```diff
@@ -19,7 +19,9 @@ public class HudHudMapLayerStore: ObservableObject {
 
     // MARK: Properties
 
-    @Published public var hudhudMapLayers: [HudHudMapLayer]?
+    @AppStorage("selectedStyleOptions") public var styleOption: MapStyleOption = []
+
+    @Published public var mapStyles: [HudHudMapStyle]?
     @Published public var lastError: Error?
```
Comment (Patrick-Kladek) [inline BackendService/Sources/BackendService/Toursprung/HudHudMapLayerStore.swift:25]: I've unified the previously two variables. It turns out that @AppStorage supports OptionSet, so there was no need for the `Int` workaround.

File: HudHud/Model/MapStore.swift [hunk 3]
```diff
@@ -160,6 +166,14 @@ final class MapStore {
 
     // MARK: Functions
 
+    func updateTrafficLayerVisibility(isVisible: Bool) {
+        guard let layers = self.mapStyle?.layers else { return }
+
+        for layer in layers where layer.matches(styleOption: .traffic) {
+            layer.isVisible = isVisible
+        }
```
Comment (Patrick-Kladek) [inline HudHud/Model/MapStore.swift:174]: adding an extension to MLNMapLayer allows us to use the where clause with a nice readable syntax.

File: BackendService/Sources/BackendService/Toursprung/HudHudMapLayerStore.swift [hunk 4]
```diff
@@ -58,11 +65,71 @@ public class HudHudMapLayerStore: ObservableObject {
...
+    }
+
+    public var layerIdentifier: String? {
+        switch self {
+        case .traffic:
+            return "traffic-lines"
+        default:
+            return nil
+        }
+    }
```
Comment (Patrick-Kladek) [inline BackendService/Sources/BackendService/Toursprung/HudHudMapLayerStore.swift:120]: I've moved this from `MapLayerIdentifier`. I think it makes sense to keep the toggle able layers here, so they can match with the buttons in the UI.

**Review by Patrick-Kladek — APPROVED** [review]
I've renamed some Models to better describe what they do, now we have:
* `MapStyle` instead of `HudHudMapLayer` - that was confusing, a style defines the layers.
* `MapStyleOption` instead of `MapStyleType`

Anything you spot wrong with my proposed changes?

**Comment by Patrick-Kladek** [comment]
Just realised, I "fixed" the size of the map style tiles but I'm using iPhone 16 Pro, your screenshots use iPhone 16 Pro Max. I guess on the max the tiles might be the wrong size. Can you try to look into this?

It might be because we use dynamic size on the sheet (1/3 height) and this is hard to translate to the tiles

File: BackendService/Sources/BackendService/Toursprung/HudHudMapLayerStore.swift (hunk 1 above)
Comment (FatooAJ) [reply BackendService/Sources/BackendService/Toursprung/HudHudMapLayerStore.swift:90]: Naif idea 🎉 he explained to me yesterday in a call how to do it.

**Comment by FatooAJ** [comment]
> Just realised, I "fixed" the size of the map style tiles but I'm using iPhone 16 Pro, your screenshots use iPhone 16 Pro Max. I guess on the max the tiles might be the wrong size. Can you try to look into this?
> 
> It might be because we use dynamic size on the sheet (1/3 height) and this is hard to translate to the tiles

looks good 

Simulator Screenshot - iPhone 16 Pro Max - 2024-11-25 at 11 30 08

**Pull request merged**